packages/lodestar/src/chain/validation/attestation.ts

```diff
@@ -1,177 +1,286 @@
 
 export async function validateGossipAttestation(
-  config: IBeaconConfig,
   chain: IBeaconChain,
-  db: IBeaconDb,
-  attestationJob: IAttestationJob,
-  subnet: number
-): Promise<phase0.IndexedAttestation> {
-  const attestation = attestationJob.attestation;
-  const numBits = getAttestationAttesterCount(attestation);
-  if (numBits !== 1) {
-    throw new AttestationError({
-      code: AttestationErrorCode.NOT_EXACTLY_ONE_AGGREGATION_BIT_SET,
-      numBits,
-      job: attestationJob,
-    });
-  }
-
-  const latestPermissibleSlot = chain.clock.currentSlot;
-  const earliestPermissibleSlot = chain.clock.currentSlot - ATTESTATION_PROPAGATION_SLOT_RANGE;
-  const attestationSlot = attestation.data.slot;
-  if (attestationSlot < earliestPermissibleSlot) {
-    throw new AttestationError({
-      code: AttestationErrorCode.PAST_SLOT,
-      earliestPermissibleSlot,
-      attestationSlot,
-      job: attestationJob,
-    });
-  }
+  attestation: phase0.Attestation,
+  /** Optional, to allow verifying attestations through API with unknown subnet */
+  subnet: number | null
+): Promise<{indexedAttestation: phase0.IndexedAttestation; subnet: number}> {
+  // Do checks in this order:
+  // - do early checks (w/o indexed attestation)
+  // - > obtain indexed attestation and committes per slot
+  // - do middle checks w/ indexed attestation
+  // - > verify signature
+  // - do late checks w/ a valid signature
 
-  if (attestationSlot > latestPermissibleSlot) {
-    throw new AttestationError({
-      code: AttestationErrorCode.FUTURE_SLOT,
-      latestPermissibleSlot,
-      attestationSlot,
-      job: attestationJob,
-    });
-  }
+  // verify_early_checks
+  // Run the checks that happen before an indexed attestation is constructed.
+  const attData = attestation.data;
+  const attSlot = attData.slot;
+  const attEpoch = computeEpochAtSlot(attSlot);
+  const attTarget = attData.target;
+  const targetEpoch = attTarget.epoch;
 
-  // no other validator attestation for same target epoch has been seen
-  if (db.seenAttestationCache.hasCommitteeAttestation(attestation)) {
+  // [REJECT] The attestation's epoch matches its target -- i.e. attestation.data.target.epoch == compute_epoch_at_slot(attestation.data.slot)
+  if (!ssz.Epoch.equals(targetEpoch, attEpoch)) {
     throw new AttestationError({
-      code: AttestationErrorCode.ATTESTATION_ALREADY_KNOWN,
-      root: ssz.phase0.Attestation.hashTreeRoot(attestation),
-      job: attestationJob,
+      code: AttestationErrorCode.BAD_TARGET_EPOCH,
     });
   }
 
-  // No need to check if `attestation.data.beaconBlockRoot` is a badBlock
-  // If it is, the ForkChoice won't know about it so the `chain.forkChoice.hasBlock` is sufficient
+  // Ensure attestation is within the last ATTESTATION_PROPAGATION_SLOT_RANGE slots (within a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+  //
+  // TODO (LH): Do not queue future attestations for later processing. Investigate why LH doesn't do it
```

Comment:
I think this is because they have `toleranceSec` check per slot range validation, that should be enough (far future slot should be invalid)

Reply:
so we can remove this 2 `TODO` concern